File: BOJ/16500_1.py
'''
    문자열 판별
    https://www.acmicpc.net/problem/16500
'''
from sys import stdin
input = stdin.readline
MAX = 100 + 1
S = input().strip()
A = []
cache = [-1] * MAX

# dp로 문자열 위치들이 될 수 있는지 없는지 검사


def search(pos):
    # 기저 : 전에 검색 완료한 경우
    if(cache[pos] != -1):
        return cache[pos]
    # pos이 S의 길이와 같은 경우
    if(pos == len(S)):
        cache[pos] = 1
        return cache[pos]
    cache[pos] = 0
    for s in A:
        M = S[pos:pos + len(s)]
        if M == s:
            cache[pos] |= search(pos + len(s))
    return cache[pos]


for _ in range(int(input())):
    A.append(input().strip())

print(search(0))

File: BOJ/16500_1.cpp
// 문자열 판별
// https://www.acmicpc.net/problem/16500
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;
const int MAX = 100 + 1;
int N;
vector<string> A;
string S;
int cache[MAX];
// 시작위치를 받아서 true, false 체크
bool classify(int pos)
{
    int &ret = cache[pos];
    if (ret != -1)
        return ret;
    if (pos == S.size()) // 정답인 경우
        return ret = 1;
    ret = 0;
    for (int i = 0; i < N; i++)
    {
        string M = S.substr(pos, A[i].size());
        if (A[i] == M)
        {
            ret |= classify(pos + A[i].size());
        }
    }
    return ret;
}
void init()
{
    fill_n(cache, MAX, -1);
    cin >> S >> N;
    A.resize(N);
    for (int i = 0; i < N; i++)
        cin >> A[i];
}
int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    init();
    cout << classify(0);

    return 0;
}
